### PR TITLE
server : support multi-modal context checkpoints

### LINKED ISSUE #19690: Eval bug:  qwen35moe always forces a full prompt reprocess after each message, 'failed to truncate'
### Name and Version

version: 8070 (cc45f2ada)
built with GNU 13.3.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

CUDA

### Hardware

Ryzen 9950x3d + 5090

### Models

The issue presents with several quants of
Qwen3.5 397B

### Problem description & steps to reproduce

When a generation is initiated, regardless if from the llama.cpp webui, sillytavern or other frontends, every generation regardless if it's a new reply or a swipe forces a full prompt reprocessing.

launch arguments: 
>/llama-server -m model/Qwen3.5-397B-A17B-UD-Q5_K_XL-00001-of-00007.gguf -ngl 999 --threads 16 --threads-batch 16 --batch-size 2048 -ub 2048 -ot "blk\.(0|1|2)\.ffn_.*=CUDA0" -ot "blk\..*_exps\.=CPU" --ctx-size 96000 --port 15000 --chat-template-kwargs "{\"enable_thinking\": false}" --mmproj Desktop/model/AI/LLama.cpp/mmproj-F32.gguf --no-mmap

### First Bad Commit

It is present since the commit adding support for qwen3.5moe

### Relevant log output

>slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 1.000 (> 0.100 thold), f_keep = 0.838
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> logit-bias -> top-k -> min-p -> ?temp-ext -> adaptive-p 
slot launch_slot_: id  3 | task 427 | processing task, is_child = 0
slot update_slots: id  3 | task 427 | new prompt, n_ctx_slot = 96000, n_keep = 0, task.n_tokens = 2189
slot update_slots: id  3 | task 427 | need to evaluate at least 1 token for each active slot (n_past = 2189, task.n_tokens() = 2189)
slot update_slots: id  3 | task 427 | n_past was set to 2188
slot update_slots: id  3 | task 427 | n_tokens = 2188, memory_seq_rm [2188, end)
slot update_slots: id  3 | task 427 | failed to truncate tokens with position >= 2188 - clearing the memory
slot prompt_clear: id  3 | task 427 | clearing prompt with 2188 tokens

### LINKED ISSUE #19858: Eval bug: Qwen3.5 always re-processes the full prompt
### Name and Version

build: 8140 (https://github.com/ggml-org/llama.cpp/commit/39fb81f8756208d79dd12ad89359f3b31675e20d) with GNU 15.2.0 for Linux x86_64

### Operating systems

Linux

### GGML backends

Vulkan

### Hardware

RX 7900XTX

### Models

Qwen3.5 Q4_K_XL Unsloth GGUF

### Problem description & steps to reproduce

I am running llama-server in docker in router mode. When I load Qwen3.5 I see the below logs which show it is not re-using the context. I set the below config so there should be plenty of context for it

```
; Context Size
ctx-size = 100000
parallel = 4 
slot-prompt-similarity = 0.25
```

### First Bad Commit

_No response_

### Relevant log output

<details>
<summary>Logs</summary>

```console
srv  proxy_reques: proxying request to model Qwen3.5 on port 53409
[53409] srv  params_from_: Chat format: peg-constructed
[53409] slot get_availabl: id  3 | task -1 | selected slot by LCP similarity, sim_best = 0.988 (> 0.250 thold), f_keep = 0.940
[53409] slot launch_slot_: id  3 | task -1 | sampler chain: logits -> penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> ?min-p -> ?xtc -> ?temp-ext -> dist 
[53409] slot launch_slot_: id  3 | task 2400 | processing task, is_child = 0
[53409] slot update_slots: id  3 | task 2400 | new prompt, n_ctx_slot = 25088, n_keep = 0, task.n_tokens = 7238
[53409] slot update_slots: id  3 | task 2400 | erased invalidated context checkpoint (pos_min = 7536, pos_max = 7536, n_swa = 1, size = 62.813 MiB)
[53409] slot update_slots: id  3 | task 2400 | n_tokens = 7153, memory_seq_rm [7153, end)
[53409] slot update_slots: id  3 | task 2400 | failed to truncate tokens with position >= 7153 - clearing the memory
[53409] slot prompt_clear: id  3 | task 2400 | clearing prompt with 7153 tokens
[53409] slot update_slots: id  3 | task 2400 | prompt processing progress, n_tokens = 2048, batch.n_tokens = 2048, progress = 0.282951
[53409] slot update_slots: id  3 | task 2400 | n_tokens = 2048, memory_seq_rm [2048, end)
[53409] slot update_slots: id  3 | task 2400 | prompt processing progress, n_tokens = 4096, batch.n_tokens = 2048, progress = 0.565902
[53409] slot update_slots: id  3 | task 2400 | n_tokens = 4096, memory_seq_rm [4096, end)
[53409] slot update_slots: id  3 | task 2400 | prompt processing progress, n_tokens = 6144, batch.n_tokens = 2048, progress = 0.848853
[53409] slot update_slots: id  3 | task 2400 | n_tokens = 6144, memory_seq_rm [6144, end)
[53409] slot update_slots: id  3 | task 2400 | prompt processing progress, n_tokens = 6726, batch.n_tokens = 582, progress = 0.929262
[53409] slot update_slots: id  3 | task 2400 | n_tokens = 6726, memory_seq_rm [6726, end)
[53409] slot update_slots: id  3 | task 2400 | prompt processing progress, n_tokens = 7238, batch.n_tokens = 512, progress = 1.000000
[53409] slot update_slots: id  3 | task 2400 | prompt done, n_tokens = 7238, batch.n_tokens = 512
[53409] slot init_sampler: id  3 | task 2400 | init sampler, took 2.27 ms, tokens: text = 7238, total = 7238
[53409] srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 200
srv  log_server_r: done request: POST /v1/chat/completions 192.168.50.105 200
[53409] slot print_timing: id  3 | task 2400 | 
[53409] prompt eval time =    3815.79 ms /  7238 tokens (    0.53 ms per token,  1896.85 tokens per second)
[53409]        eval time =      15.81 ms /     2 tokens (    7.90 ms per token,   126.53 tokens per second)
[53409]       total time =    3831.60 ms /  7240 tokens
[53409] slot      release: id  3 | task 2400 | stop processing: n_tokens = 7239, truncated = 0
[53409] srv  update_slots: all slots are idle
```
</details>

### PR DESCRIPTION
**Pull request by ggerganov**
alt #19747 
fix #19690
fix #19858

This continues the work from #19747 to enable multi-modal checkpointing in `llama-server`. Avoids changes to `libllama`. Fixes `n_past` ambiguity of "number of tokens" and "next position" by tracking the position separately with a new `pos_next` var.

### CODE REVIEW

**Comment by timkhronos** [comment]
I think it might make more sense to post this here instead of the original pr:

@ggerganov These 3 changes should fix the incorrect checkpoint reuse I mentioned in regards to   #19849 

in server-task.h, add a new pos_end to checkpoint struct:
```
struct server_prompt_checkpoint {
    llama_pos pos_min;
    llama_pos pos_max;
    llama_pos pos_end; //full prompt extent when checkpoint was created

    std::vector<uint8_t> data;
```
in sever-context.cpp at checkpoint creation:

```
auto & cur = slot.prompt.checkpoints.emplace_back(server_prompt_checkpoint{
    /*.pos_min = */ pos_min,
    /*.pos_max = */ pos_max,
    /*.pos_end = */ slot.prompt.tokens.pos_next(),
    /*.data   = */ std::vector<uint8_t>(checkpoint_size),
});
```

and also in server-context.cpp, when checking if we have valid checkpoints:
```
const llama_pos prompt_end = slot.task->tokens.pos_next();

const auto it = std::find_if(
    slot.prompt.checkpoints.rbegin(),
    slot.prompt.checkpoints.rend(),
    [&](const auto & cur) {
        if (cur.pos_end > prompt_end) {
            return false;
        }
        // guarantee that a checkpoint will result in at least one token being processed [TAG_PROMPT_LOGITS]
        return cur.pos_min < pos_min_thold;
    }
);
```

**Comment by ggerganov** [comment]
@ngxson PTAL, I think this is good to merge.

File: tools/server/server-task.h [hunk 1]
```diff
@@ -557,6 +557,8 @@ struct server_prompt_checkpoint {
     llama_pos pos_min;
     llama_pos pos_max;
 
+    int n_tokens;
```
Comment (ngxson) [inline tools/server/server-task.h]: should we change this to `int64_t` to match what's `pos_next(int64_t)` expects?

**Review by ngxson — APPROVED** [review]
